### DISCUSSION TITLE
Объединение плюсов fsh и zendoc

### DISCUSSION THREAD

**Discussion by eugenekogan-spb** [fhir-ru/core#224] [General]
Коллеги, поскольку на [встрече 14 февраля](http://fhir-ru.zendoc.me/meetings.2023-02-14) мы замахнулись на объединение плюсов стандартных инструментов и плюсов zendoc, предлагаю сформулировать это конкретно.

Плюсы стандартных инструментов, которые хочется получить:
- стандартное размещение для стандартных элементов руководства (схема профиля, детальное описание атрибутов профиля, примеры)
- автоматически (из одного исходника) ставятся ссылки между артефактами – профиль, набор значений, система кодирования
- ссылки между страничками гарантированно работают
- работающая валидация схемы профиля
- работающая валидация экземпляров ресурсов
- возможность генерации машиночитаемого комплекта Implementation Guide, как одного из результатов работы
- мелкие приятности, например, качественная отрисовка дифференциальной схемы (что именно изменилось)
- полная документация

Плюсы zendoc, которые хочется не потерять:
- «википедия», в которой отражена работа сообщества, есть записи встреч, методические материалы, различные руководства по реализации, и эти материалы собраны на одном сайте;
- руководство получается русскоязычным;
- добавление своих функций, например валидация экземпляров онлайн, XSLT-преобразователь онлайн, потенциально валидация по наборам значений и взаимодействие со справочниками МЗРФ;
- добавление к стандартной спецификации объясняющих страниц, например страниц с логическими схемами, со схемами процессов, которые сейчас делает клиническая группа, для того чтобы показать, почему были приняты те или иные решения;
- возможность программно обращаться к данным на страницах (дай все профили пациента), которая сейчас не задействована, но на будущее представляется ценной.

Предложенное решение:

Этап 1
- накатить новую версию zendoc, в ней есть онлайн-редактирование и страница проверки ссылок
- сделать возможность встраивать текст на fsh на страницы zendoc, при валидации движок будет сам переводить эти тексты в edn
- доделать валидацию схем и валидацию экземпляров 
- сделать "верхние табы" для размещения Content, Detailed Description и Examples на страничке профиля

Этап 2
Если этап 1 успешен, продолжать разработку RuCore и связанных руководств, и параллельно определить приоритеты дальнейших инфраструктурных работ.

**Comment by eugenekogan-spb** [comment]
@Vrodvi @OlegPenzin @niquola прошу поправить-добавить

**Reply by Vrodvi** [reply]
Евгений, большое спасибо за тщательно проведённую систематизацию требований.

Хотелось бы добавить ещё несколько положительных моментов от использования стандартных инструментов, которые также хотелось бы привнести в zendoc:

- снижение порога входа для составителей профилей за счёт упрощённого синтаксиса (DSL ) описания профилей ([FHIR Shorthand](http://build.fhir.org/ig/HL7/fhir-shorthand/reference.html)) с большим количеством готовых примеров ([fsh-finder](https://fshschool.org/fsh-finder/)) и онлайн-инструментом для конвертации в стандартные StructureDefinition ([конвертер](https://fshschool.org/FSHOnline/#/))
![image](https://user-images.githubusercontent.com/92928137/219406231-cdbe98e5-e7c7-4790-b0d1-9add78ba9fc9.png)

- возможность скачать описание профиля в табличном представлении (csv, excel)
![image](https://user-images.githubusercontent.com/92928137/219401342-66a0aeb3-fb03-40b4-9635-dc64e0d0843b.png)

- автоматическая генерация человекочитаемых представлений экземпляров ресурсов (например, https://build.fhir.org/ig/HL7/fhir-mCODE-ig/Patient-cancer-patient-adam-everyman.html)
(пусть и не идеальная, но она есть, и для ознакомления с наполнением подходит лучше чем просмотр сырого JSON) 
![image](https://user-images.githubusercontent.com/92928137/219405849-e4de7bfa-f093-4fec-8d95-e892b8db4d00.png)

- автоматическое развёртывание наборов значений в человекочитаемые представления
![image](https://user-images.githubusercontent.com/92928137/219405668-2b8a104f-3af8-42a4-9512-f5fc9f63e1bd.png)
  - в том числе при развёртывании набора значений учитываются и отрисовываются все связанные с ним маппирования на другие наборы значений (ConceptMap)


Про пункт "возможность генерации машиночитаемого комплекта Implementation Guide, как одного из результатов работы" хотелось бы расширить:
- "... с автоматической генерацией сайта руководства по реализации стандартными средствами IG Publisher"
  - в том числе это нужно для ведения версионности руководств по реализации, как например тут - http://hl7.org/fhir/us/mcode/history.html 
![image](https://user-images.githubusercontent.com/92928137/219408535-b239dd0b-d332-42f0-bb65-67cd37b8bfd7.png)

**Comment by Vrodvi** [comment]
@eugenekogan-spb @OlegPenzin @niquola 

Провели первое собрание с Николаем по поводу походов к объединению zendoc и fsh (с сопутствующими технологиями).

Обсудили несколько подходов:

1) Ведение zendoc только как базы знаний со страницами с общими сведениями, а все FHIR артефакты вести в отдельном проекте и генерировать отдельный сайт руководства по реализации. Далее давать ссылку из Zendoc'а на сайт руководства.
[НЕ ПРИНЯТО] Т.к. теряются преимущества Zendoc и для читателей неудобно бегать между двумя сайтами.

2) Встраивание fsh в страницы Zendoc с полным повторением всей функциональности стандартных сайтов руководств по реализации, но средствами Zendoc.
 [НЕ ПРИНЯТО] Т.к. объём стандартной функциональности слишком большой и повторить её в полном объёме с приемлемым качеством будет проблематично. Также остаётся проблема в нестандартном визуальном представлении наполнения руководства.

3) Встраивание fsh в страницы Zendoc только с частичным повторением нужной нам функциональности стандартных сайтов руководств по реализации. Параллельно сделать сборку всех фрагментов fsh, разбросанных по Zendoc'у, и генерацию на их основе полностью стандартного сайта руководства по реализации, к которому можно получить отдельный доступ.
[ПРИНЯТО] Т.к. позволит оставить гибкость описания Zendoc и в то же время будет обеспечена генерация дополнительного сайта со стандартной структурой.


Для демонстрации жизнеспособности принятого подхода решено составить в Zendoc фрагмент документации с двумя страницами с fsh описаниями двух взаимосвязанных профилей, составить скрипт их сборки и генерации на их основе стандартного сайта с руководством по реализации.
Для примера возьмём два профиля из клинической группы, составленных в формате fsh (сами профили приложим в эту дискуссию чуть позже).

Дополнительные сведения и материалы по ходу указанных работ предлагаю и дальше публиковать в этой дискуссии.

**Reply by Vrodvi** [reply]
@eugenekogan-spb @OlegPenzin @niquola 

Два профиля (со всеми сопутствующими описаниями - CodeSystem, ValueSet, Extension и т.п.) подготовлены.

Как было оговорено, профили разделены на два файла (при этом файлы содержат перекрёстные ссылки): 

- Для страницы с описанием Encounter: **core-example-part-1-encounter-resources.fsh** ([core-example-part-1-encounter-resources.zip](https://github.com/fhir-ru/core/files/10852665/core-example-part-1-encounter-resources.zip))

- Для страницы с описанием Condition: **core-example-part-2-condition-resources.fsh** ([core-example-part-2-condition-resources.zip](https://github.com/fhir-ru/core/files/10852668/core-example-part-2-condition-resources.zip))

Инструментом SUSHI данные файлы успешно конвертируются и строится сайт руководства по реализации.
[image](https://user-images.githubusercontent.com/92928137/221917887-5fbe6837-10bd-4c9c-9842-be85ab823562.png)

Онлайн-инструментом при склеивании в один текстовый массив также успешно конвертируются:
[image](https://user-images.githubusercontent.com/92928137/221905445-b15ecff3-9eb9-4a2b-b24d-f4022377b179.png)

**Reply by Vrodvi** [reply]
@eugenekogan-spb 
Евгений, если есть люди, готовые начать разбираться со стандартным инструментарием, то можно, параллельно с работами по встраиванию редактора FSH коллегами из HealthSamurai, пробовать формировать IG через инструмент SUSHI напрямую.

Как раз будет возможность сравнить результат по созданию IG напрямую из SUSHI и опосредованно через Zendoc.

Для работы с SUSHI необходимо установить и настроить среду разработки как описано в инструкции https://fshschool.org/quickstart/ (если нужна инструкция по настройке на русском языке, то можно запустить работу по её подготовке в разделе "Методические материалы" на сайте Zendoc).

Альтернативно можно попробовать использовать пред-настроенную среду разработки в виде Docker-контейнера, подготовленную коллегами в общем чате сообщества (https://t.me/fhir_ru/4069).

Если заведём страницу, то можно там оба вариант указать - на выбор.
В любом случае хотя бы небольшую заготовку для инструкции нужно завести сейчас и затем дополнять её по мере возможностей.

В качестве входных данных для генерации IG через SUSHI можно взять представленные выше материалы:
core-example-part-1-encounter-resources.fsh
core-example-part-2-condition-resources.fsh

**Reply by Vrodvi** [reply]
> Альтернативно можно попробовать использовать пред-настроенную среду разработки в виде Docker-контейнера, подготовленную коллегами в общем чате сообщества (https://t.me/fhir_ru/4069).

@eugenekogan-spb @OlegPenzin 
Коллеги мне кажется вариант с преднастроенной средой в Docker-контейнере будет более простым и поддерживаемым. Но даже для него необходимо написать хотя бы краткую инструкцию по использованию. Если есть возможность, просьба рассмотреть возможность привлечения коллег, подготовивших Docker-контейнер (https://t.me/fhir_ru/4069) к работе по написанию инструкции.

**Reply by Vrodvi** [reply]
Пока описание по работе с преднастроенной средой в Docker-контейнере не подготовлено, возникли неожиданные технические затруднения. Последний комментарий по настройке Docker-контейнара можно посмотреть по ссылке - https://t.me/fhir_ru/4126.

Как обсуждали - альтернативно подготовлена краткая инструкция по работе через онлайн-инструмент - http://fhir-ru.zendoc.me/support.instr#:onlineFsh
Описать упомянутый в этой краткой инструкции валидатор и публикатор пока что не было возможности - для этого нужно подобрать самый простой вариант использования, проверить его работоспособность и описать именно его, т.к. исходная документация по этим инструментам слишком обширна со множеством конфигураций, не все из которых нам нужны.
